magma/primitives/when.py

```diff
@@ -203,16 +207,22 @@ def add_default_driver(self, drivee: Type, driver: Type):
         self._default_drivers[drivee] = driver
 
     def _finalize(self):
+        # Detect latches which would be inferred from the context of the when
+        # block.
+        latches = find_inferred_latches(self.block)
+
         # NOTE(rsetaluri): These passes should ideally be done after circuit
         # creation. However, it is quite unwieldy, so we opt to do it in this
         # builder's finalization, although it is a bit "hacky". Ideally, this
         # builder should be completely agnostic of anything it is connected to,
         # and a global (post-processing) pass should be used instead.
-        _add_default_drivers_to_memory_ports(self)
-        _add_default_drivers_to_register_inputs(self)
-        # Detect latches which would be inferred from the context of the when
-        # block.
-        latches = find_inferred_latches(self.block)
+        # NOTE(leonardt): we only add default drivers if a value is inferred to
+        # be a latch, otherwise we skip this implicit logic so the output is
+        # cleaner (e.g. if a register is assigned in every case, we don't need
+        # the default value)
```

Comment:
```suggestion
        # NOTE(leonardt): We only add default drivers if a value is inferred to
        # be a latch, otherwise we skip this implicit logic so the output is
        # cleaner (e.g. if a register is assigned in every case, we don't need
        # the default value).
```